pkg/finality-grandpa/voting_round.go

```diff
@@ -0,0 +1,849 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+	"time"
+
+	"golang.org/x/exp/constraints"
+)
+
+type stateStart[T any] [2]T
+
+type stateProposed[T any] [2]T
+
+type statePrevoting[T, W any] struct {
+	T T
+	W W
+}
+
+type statePrevoted[T any] [1]T
+
+type statePrecommitted struct{}
+
+// type states[T, W any] interface {
```

Comment:
nit/remove

Reply:
ended up using it, and creating `setState` and `newState` functions to use this interface constraint.